### PR TITLE
Feature kinesis forwarder

### PR DESCRIPTION
**Pull request by zabullet**
Initial spike at integrating AWS Kinesis in to this tool.

Still need to do test, but fundamentally this is an RFC

### CODE REVIEW

File: kinesis/forwarder.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+	name            string
+	kinesisClient   kinesisiface.KinesisAPI
+	streamName      string
+	outputQ         *[]*kinesis.PutRecordsRequestEntry
+	lastSuccessTime *int64
+}
+
+// CreateForwarder creates instance of forwarder
+func CreateForwarder(entry config.AmazonEntry, kinesisClient ...kinesisiface.KinesisAPI) forwarder.Client {
+	log.Info("Test")
```
Comment (filiphaftek) [inline kinesis/forwarder.go]: What is this line? :)

File: kinesis/forwarder.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+	inputRecord := &kinesis.PutRecordsRequestEntry{
+		Data:         []byte(message),                            // Required
+		PartitionKey: aws.String(strconv.Itoa(rand.Intn(10000)))} //use something random for partition
+
+	*f.outputQ = append(*f.outputQ, inputRecord)
+
+	currentUnixTime := time.Now().UnixNano()
+
+	if (currentUnixTime-*f.lastSuccessTime >= 1000*1000*1000) ||
```
Comment (filiphaftek) [inline kinesis/forwarder.go]: Why so weird conversion? Why do you assume 1 seconds and it's not configurable?

File: kinesis/forwarder.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+		(len(*f.outputQ) >= 500) { //maximize for 1 second bursts or 500 items
```
Comment (filiphaftek) [inline kinesis/forwarder.go]: Why 500 items hardcoded?

File: kinesis/forwarder.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+		f.flushQueuedMessages()
+	}
+
+	return nil
+}
+
+// Stop stops the forwarder in this case it attempts a flush
+func (f Forwarder) Stop() error {
+	log.Info("Stopping")
```
Comment (filiphaftek) [inline kinesis/forwarder.go]: `Stopping` is not very descriptive, would appreciate more information in the sentence i.e. forwarderName etc :)

File: supervisor/supervisor.go [hunk 5]
```diff
@@ -97,6 +98,12 @@ func (c *Client) Restart(w http.ResponseWriter, r *http.Request) {
 	successResponse(w)
 }
 
+// Stop stops the
```
Comment (filiphaftek) [inline supervisor/supervisor.go]: `stops the ...` ?

File: kinesis/forwarder.go (hunk 2 above)
Comment (zabullet) [reply kinesis/forwarder.go]: This section needs some work. There is no good justification for the timing and hard coded values.

The problem I'm attempting to tackle and balance is keeping the message rate high without triggering rate limits on kinesis. A single shard has a 1K tps limit before starting to fail the call.

...anyway.....this implementation doesn't really address this in any way that is remotely robust, just badly avoids it.

I'll rework it with a proper backoff mechanism

**Comment by zabullet** [comment]
Thanks for the feedback. All fair comments. I'll do some clean-up and add tests before updating PR.

File: kinesis/forwarder.go (hunk 1 above)
Comment (zabullet) [reply kinesis/forwarder.go]: Removed

**Comment by zabullet** [comment]
Ignore. Forgot something.

File: kinesis/forwarder.go (hunk 2 above)
Comment (zabullet) [reply kinesis/forwarder.go]: Fixed conversion and added parameter

File: kinesis/forwarder.go (hunk 3 above)
Comment (zabullet) [reply kinesis/forwarder.go]: 500 is the maximum that you can batch with the Kinesis API.....I could let it get bigger and then batch in 500s, but I'm not sure what advantage that would have.

File: kinesis/forwarder.go (hunk 4 above)
Comment (zabullet) [reply kinesis/forwarder.go]: Improved comment

File: supervisor/supervisor.go (hunk 5 above)
Comment (zabullet) [reply supervisor/supervisor.go]: Update comment

**Comment by zabullet** [comment]
I think this is a better PR. Please review and let me know anything else you want changed.

**Comment by zabullet** [comment]
Closing PR. I've done a bunch of work on a development branch, which moves on from this. I'll consolidate and do a new PR

**Pull request closed without merging**